File: src/run_sync_client.cpp
#include "sync_client.h"

#include <string>
#include <sstream>

int main(int argc, char* argv[] ) {

    std::string addressStr = "localhost";
    std::string portStr = "50051";

    if (argc > 1) {

        addressStr = argv[1] ;
    }

    if (argc > 2) {

        portStr = argv[2] ;
    }

    std::string target_str = addressStr + ":" + portStr; 

    GeoClient client( ::grpc::CreateChannel(target_str, grpc::InsecureChannelCredentials() ) ) ;

    std::string lineStr;

    std::cout << "Supported commands: ADD, ADDMANY, REMOVE, GETALL, GETBYCOORDS, GETBYNAME" << std::endl;

    while ( std::getline(std::cin, lineStr) ) {

        std::stringstream ss(lineStr);

        std::string command;
        ss >> command;

        if (command == "ADD") {

            GeoHelper::CityInfo cityInfo;

            ss >> cityInfo.name >> cityInfo.point.latitude >> cityInfo.point.longitude;

            bool result = client.AddCity(cityInfo) ;

            std::cout << std::boolalpha << result << std::endl;
        }

        else if (command == "REMOVE") {

            GeoHelper::CityInfo cityInfo;
            
            ss >> cityInfo.name >> cityInfo.point.latitude >> cityInfo.point.longitude;

            bool result = client.RemoveCity(cityInfo) ;

            std::cout << std::boolalpha << result << std::endl;

        }

        else if (command == "GETALL" ) {

            auto result = client.GetCities() ;

            for (const auto& item : result) {

                std::cout << item.name << " " << item.point.latitude << " : " << item.point.longitude << std::endl;
            }
        }

        else if (command == "GETBYCOORDS") {


            GeoHelper::Point point;
            
            ss >> point.latitude >> point.longitude;

            auto result = client.GetCityByCoords(point) ;

            if (!result.empty()) {

                std::cout << "Found city " << result << " at " << point.latitude << ":" << point.longitude << std::endl;
            }
            else {

                std::cout << "City not found" << std::endl;
            }

        }
        else if (command == "GETBYNAME") {


            std::string cityName;
            ss >> cityName;

            auto result = client.GetCityByName(cityName) ;

            for (const auto& point : result) {

                std::cout << "Found city " << cityName << " at " << point.latitude << " : " << point.longitude << std::endl;
            }


        }

        else if (command == "ADDMANY" ) {

            int count;
            ss >> count;

            std::vector<GeoHelper::CityInfo> input;
            input.reserve(count);

            for (int i = 0; i < count; ++i) {

                GeoHelper::CityInfo cityInfo;

                ss >> cityInfo.name >> cityInfo.point.latitude >> cityInfo.point.longitude ;

                input.push_back(cityInfo);
            }

            auto result = client.AddCities(input);

            for (int i = 0; i < result.size() ; ++i) {

                std::cout << std::boolalpha << result[i] << std::endl;
            }
        }

        else {

            std::cout << "Unknown command" << std::endl;
        }
    }

    return 0;
}

File: src_py/run_sync_client.py
import grpc

import geo_pb2
import geo_pb2_grpc




def queryAllCities(stub):
      
    response = stub.GetCities(geo_pb2.Void())

    for item in response:
        print(item.name)
        print(item.location)

def addCity(stub):

    response = stub.AddCity(geo_pb2.CityInfo(name="Moscow", location=geo_pb2.Point(latitude=3, longitude=9)))
    print(response.status)


def removeCity(stub):

    response = stub.RemoveCity(geo_pb2.CityInfo(name="Moscow", location=geo_pb2.Point(latitude=3, longitude=9)))
    print(response.status)


def generateCityInfoList():

    cities = [ geo_pb2.CityInfo(name="Samara", location=geo_pb2.Point(latitude=23, longitude=19)) ,
            geo_pb2.CityInfo(name="Saratov", location=geo_pb2.Point(latitude=253, longitude=1)),
            geo_pb2.CityInfo(name="Sarov", location=geo_pb2.Point(latitude=2223, longitude=191))
    ]

    for city in cities:
        yield city
    
def addCities(stub):

    responses = stub.AddCities(generateCityInfoList() )

    for response in responses:
        print(response.status)

def getByName(stub):

    responses = stub.GetCityByName( geo_pb2.CityName(name="Moscow") )

    for response in responses:
        print(response)


def getByCoords(stub : geo_pb2_grpc.GeoManagerStub):

    response = stub.GetCityByCoords(geo_pb2.Point(latitude=23, longitude=19) )
    print(response.name)


def run():

    print("Client starts working ...")

    with grpc.insecure_channel('localhost:50051') as channel:
        
        stub = geo_pb2_grpc.GeoManagerStub(channel)

        queryAllCities(stub)
        addCity(stub)
        removeCity(stub)
        addCities(stub)
        getByName(stub)
        getByCoords(stub)
        
    
    print("Client finished OK")


if __name__ == '__main__':
    run()